### DISCUSSION TITLE
Parallelizing with BiocParallel

### DISCUSSION THREAD

**Discussion by llrs** [ropensci/targets#665] [Help] [answered]
I am using a function parallelized with BiocParallel on a single machine. I set up 4 cores for the target, but instead I see all the threads (8) of my computer full to 100% and the load average to 23 (which seems to indicate there are a lot of tasks pending).  

```r
 tar_target(models1, search_model(param1, param2, paramn, BPPARAM = BiocParallel::MulticoreParam(workers = 4))
```
I have checked the documentation and other discussions but I'm not sure what is going on here, and didn't find any mention to BiocParallel on previous discussions. A little guidance would be appreciated. Thanks!!

**Comment by wlandau** [comment, marked as answer]
I am unfamiliar with `BiocParallel`. Do you have a reproducible example? 

`targets` creates a new background process to run the pipeline. What happens with `tar_make(callr_functIon = NULL)`?

What happens when you run `search_model(param1, param2, paramn, BPPARAM = BiocParallel::MulticoreParam(workers = 4)` without using `targets`?

**Reply by llrs** [reply]
If I run the function alone it works well, I used it previously without incidents on the same machine, OS, and packages versions. 
It might be a bit hard for me to provide a reprex for this. I've already waited 2 weeks for this process in serial and I don't want to cancel the process for further testing as I'm a bit under time pressure. I'll check if I can reproduce it on another machine but it might take some time. 
However, this raises the question, how can I graciously cancel a target? Other times I just closed the RStudio instance but then I had some targets as started and took some time to debug and be able to restart them.

**Reply by llrs** [reply]
Mmh, at the moment I'm running the same code and R version and presumably packages versions on another machine and I don't see all the cores at 100% or anything weird. 

The problem on my computer might be due to canceling tasks (by closing Rstudio) and restarting the whole `tar_make` again which might have restarted previously paused jobs that also took several cores. What is the best way to stop a target and clean the metadata? Just use `tar_invalidate` on the failed tasks or directly `tar_prune`? Sorry for the noise!

**Reply by llrs** [reply]
For completeness, even after closing RStudio and doing the following

```r
> tar_delete("models2")
> tar_invalidate("models2")
> tar_prune()
> tar_visnetwork()
```

The visualization show that the task is started: 
![targets_started](https://user-images.githubusercontent.com/6818218/138123301-ed040724-6068-4b4a-a0b1-f446603502ef.png)

But the metadata doesn't have any input for this target and the cores are back to their normal load. I hesitate to restart and run `tar_make()` to add a new 4 core process on an 8 core machine.

**Reply by wlandau** [reply]
`targets` relies on other packages like `callr`, `clustermq`, and `future` to know how to clean up child processes when the main process is canceled. In your case, that involves some operating system signal handling, which does not always work. Restarting your session in the RStudio IDE should work if you ran `tar_make()` that way. Alternatively, `targets` provides is the PID of the *inner* main process in `_targets/meta/process`. You can loop up the PID with `tar_pid()`, cancel it with `kill -9 ***`, and hope it cleans up the `BiocParallel` processes. You may have to `kill -9` the `BiocParallel` processes manually if they do not shut down on their own. (This is the sort of reason why I prefer `callr` and `processx` over stuff based on `parallel` and `snow`.)

If you stop a pipeline while it is running, you do not need to manually clear out the metadata before running again. If target `x` was in the middle of running but did not have the chance to finish, the next `tar_make()` will start it again.

**Reply by llrs** [reply]
After restarting the machine `tar_visnetwork` still shows the process as started, and the `tar_pid()` and killing it didn't find the PID. 

I explored a bit and find on `_targets/meta/`, on `_targets/meta/progress` I see several process marked as started (`grep "started" progress`):

```
b0|stem|b0|0|started
models1|stem|models1|0|started
b1|stem|b1|0|started
models2|stem|models2|0|started
```

Some of these entries are duplicated :

```
b1|stem|b1|0|started
b1|stem|b1|0|built
models1|stem|models1|0|started
models1|stem|models1|0|built
models2|stem|models2|0|started
b0|stem|b0|0|started
b0|stem|b0|0|built
```

`tar_visnetwork` shows models2 as started because there isn't a clean up when I interrupt the process (or with `tar_invalidate` and `tar_delete`). On duplicated entries it seems that if there is something else than started, that other progress is shown. When I manually delete the line `models2|stem|models2|0|started` the visualization show that target as outdated. 

About the cores used, I'm still investigating the issue: on the same machine in a different project the (Bioc)parallel system works as expected. However, it doesn't seem to be a targets issue because when I run the same target outside `tar_make()` but on the same project I see the same behavior of using all the cores. It might be due to renv or other weird issue. 
Thanks for your help!

**Reply by wlandau** [reply]
A progress of `"started"` just says the target began running at some point. It does not actually say whether the target is still running. That would require knowing something about the computer architecture and how you spawned the processes, and because `targets` works through abstractions `future` and `clustermq`, that part is harder and `targets` does not tackle it.

The duplicated lines are also fine. When you call `tar_meta()` or `tar_progress()`, only the most recently written lines are returned. The files are cleaned up at various points in time so that the files do not grow out of control.

I think these processes are best managed through `top`, `htop`, or similar tool on your local machine. It is good practice to make sure all the previous jobs are cleaned up before submitting another round. I have had problems with abandoned/zombie processes in the past using `parallel`. The signal handling is sometimes brittle, which might be the issue on your end. I suspect this is more likely to be an issue in `BiocParallel` and `parallel` than it is to be in `renv`.
